### ISSUE TITLE
Somehow exchange versions in a way that won't add too much overhead and won't break

### ISSUE THREAD

**Issue by jD91mZM2** [synac-chat/server#2]
Currently an incompatible client will work until it doesn't. Then it will start behaving weirdly and nobody will know why.

Would it be better if errors could be detected earlier on?

**Pros:**
 - It's clear when clients are compatible and when they are not.
 - Perhaps there could be a compatibility mode for newer clients and older servers. That wouldn't be possible unless it actually knows if it's compatible.

**Cons:**
 - Clients that don't make use of whatever change there was would still break.
 - If there would be an exploit in synac and the server is running an old version, an attacker would know about it. (although synac is written in Rust, so I doubt there will be one 😎)

**Comment by Mnpn** [comment]
Yes, exchange versions. Maybe make some versions backwards compatible but only as far as possible? E.g. client sends 0.2.4 and server is running 1.3.7, no problem, but e.g. client sends 0.1.0 and server is 1.8.4 (where the client doesn't even have support for roles) it would deny the connection?

That'll probably be slow, hard and messy but it's just an idea.

**Comment by jD91mZM2** [comment]
That's the entire point of semver. \<major\>.\<minor\>.\<patch\>. I don't increase major if they're not compatible.

**Comment by jD91mZM2** [comment]
I just thought of an issue: What if somebody makes their own server side application? We can't force them to use the same version numbers as us?  
Solution: We make a protocol version number instead of using the application version.

**Comment by jD91mZM2** [comment]
Another issue: How should the version even be exchanged? The protocol might change completely which means it breaks the version exchange... And unless we put it in the LoginSuccess packet we'd need to make a whole new packet for the version stuff, which would be annoying.

**Comment by jD91mZM2** [comment]
BUMP
